File: templates/new-module/module.py
"""Module class with lifecycle hooks."""

from typing import TYPE_CHECKING

from system.db.database import db
from system.module.hooks import hookimpl

if TYPE_CHECKING:
    from flask import Blueprint


class ModuleNameModule:
    """Module class implementing lifecycle hooks.

    Rename 'ModuleName' and 'modulename' to your actual module name.
    """

    def get_routes(self) -> list[tuple["Blueprint", str]]:
        """Return list of (blueprint, url_prefix) tuples.

        Called by ModuleLoader.register_routes() during app startup.
        The url_prefix should match main_route in __manifest__.py.
        """
        from .controllers.routes import blueprint

        return [(blueprint, "/modulename")]

    @hookimpl
    def init_database(self) -> None:
        """Initialize database tables and sample data.

        Called after all modules are loaded and db.create_all() runs.
        Use try/except to handle errors gracefully.
        """
        db.create_all()

        # Optional: Create sample data
        # try:
        #     from .models.item import Item
        #     Item.create_sample_data()
        # except Exception as e:
        #     import logging
        #     logging.getLogger(__name__).error(f"Sample data error: {e}")

File: blueprint-go/templates/new-module/module.go
package modulename

import (
	"yourapp/internal/modules/modulename/handlers"
	"yourapp/internal/modules/modulename/models"
	"yourapp/internal/system/db"
	"yourapp/internal/system/module"

	"gorm.io/gorm"
)

// Ensure Module implements required interfaces
var _ module.Module = (*Module)(nil)
var _ module.DatabaseInitializer = (*Module)(nil)

// Module implements the module interface for modulename
type Module struct {
	db      *gorm.DB
	handler *handlers.Handler
}

// New creates a new module instance
func New(database *gorm.DB) *Module {
	repo := models.NewItemRepository(database)
	return &Module{
		db:      database,
		handler: handlers.NewHandler(repo),
	}
}

// GetRoutes returns the routes for this module
func (m *Module) GetRoutes() []module.Route {
	return []module.Route{
		{
			Handler: m.handler.RegisterRoutes,
			Prefix:  "/modulename",
		},
	}
}

// InitDatabase initializes database tables and sample data
func (m *Module) InitDatabase() error {
	// Auto-migrate models
	if err := m.db.AutoMigrate(&models.Item{}); err != nil {
		return err
	}

	// Create sample data (idempotent)
	repo := models.NewItemRepository(m.db)
	return repo.CreateSampleData()
}

// GetManifest returns the module manifest
func (m *Module) GetManifest() ModuleManifest {
	return Manifest
}
